### PR TITLE
Fix IN expr for NaN

### LINKED ISSUE #7377: IN expr doesn't work properly for NaN
### Describe the bug

Given the following query.
```
SELECT 'NAN'::double in ('NAN'::double);
```
I expected the result is `true` but the actual is `false`.
It's not inconsistent with the result of `'NaN'::double = 'NaN'::double`, which is evaluated as `true`.

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sarutak**
## Which issue does this PR close?
Closes #7377 

## Rationale for this change
This PR fixes an issue that `'NaN'::double in ('NaN'::double)` is evaluated as `false`, which is inconsistent with the result of `'NaN'::double = 'NaN'::double`.

## What changes are included in this PR?
The root cause is that equality is evaluated using `PartialEq::eq`. So the fix is using `to_bits` for equality check for float values like as `ScalarValue::Float64` does.

NOTE: This issue doesn't happen when `Simplifier` simplifies `IN` expr like
```
a IN (x, y, z) -> a = x OR a = y OR a = z
```

## Are these changes tested?
Modified an existing test.

## Are there any user-facing changes?
Yes, but this change is for correctness.

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 1]
```diff
@@ -609,50 +643,100 @@ mod tests {
...
+            Some(0.2),
+            None,
+            Some(f64::NAN),
+            Some(-f64::NAN),
+        ]);
         let col_a = col("a", &schema)?;
         let batch = RecordBatch::try_new(Arc::new(schema.clone()), vec![Arc::new(a)])?;
 
-        // expression: "a in (0.0, 0.2)"
+        // expression: "a in (0.0, 0.1)"
```
Comment (sarutak) [inline datafusion/physical-expr/src/expressions/in_list.rs:656]: Some comments are wrong and it's tiny issue. So I'll fix them within this PR.

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 2]
```diff
@@ -141,7 +141,7 @@ where
 fn make_hash_set<T>(array: T) -> ArrayHashSet
 where
     T: ArrayAccessor,
-    T::Item: PartialEq + HashValue,
+    T::Item: IsEqual + HashValue,
```
Comment (viirya) [inline datafusion/physical-expr/src/expressions/in_list.rs:144]: ditto

**Review by viirya — COMMENTED** [review]
Can we add end-to-end test for this too?

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 3]
```diff
@@ -94,7 +94,7 @@ impl<T> Set for ArraySet<T>
 where
     T: Array + 'static,
     for<'a> &'a T: ArrayAccessor,
-    for<'a> <&'a T as ArrayAccessor>::Item: PartialEq + HashValue,
+    for<'a> <&'a T as ArrayAccessor>::Item: IsEqual + HashValue,
```
[the remark this answers is not in the document]
Comment (sarutak) [reply datafusion/physical-expr/src/expressions/in_list.rs]: Indeed. I've fixed it.

File: datafusion/sqllogictest/test_files/predicates.slt [hunk 4]
```diff
@@ -249,6 +249,40 @@ SELECT * FROM test WHERE column1 IN ('foo', 'Bar', 'fazzz')
...
+query R
+SELECT * FROM test_float WHERE column1 IN (0.0, -1.2, NULL)
+----
+
+query R
+SELECT * FROM test_float WHERE column1 IN (0.0, 1.2, 'NaN'::double, '-NaN'::double)
+----
+1.2
+NaN
+NaN
```
[the remark this answers is not in the document]
Comment (sarutak) [reply datafusion/sqllogictest/test_files/predicates.slt:285]: It's actually `-NaN` but just shown as `NaN`.
I've already noticed this issue and have plan to fix it.

File: datafusion/sqllogictest/test_files/predicates.slt [hunk 5]
```diff
@@ -249,6 +249,40 @@ SELECT * FROM test WHERE column1 IN ('foo', 'Bar', 'fazzz')
 foo
 fazzz
 
+statement ok
+CREATE TABLE IF NOT EXISTS test_float AS VALUES(1.2),(2.1),(NULL),('NaN'::double),('-NaN'::double);
+
+# async fn in_list_float
```
[the remark this answers is not in the document]
Comment (sarutak) [reply datafusion/sqllogictest/test_files/predicates.slt:255]: It's a convention isn't it?
I saw some similar test names like ` async fn in_set_string_dictionaries` and `async fn in_list_string_dictionaries`, which don't have the corresponding functions.

File: datafusion/sqllogictest/test_files/predicates.slt (hunk 4 above)
[the remark this answers is not in the document]
Comment (viirya) [reply datafusion/sqllogictest/test_files/predicates.slt:285]: Okay

File: datafusion/sqllogictest/test_files/predicates.slt (hunk 5 above)
[the remark this answers is not in the document]
Comment (sarutak) [reply datafusion/sqllogictest/test_files/predicates.slt:255]: O.K, I understand. I'll modify the name for the new test.

**Review by alamb — APPROVED** [review]
Thank you @sarutak 

cc @tustvold  for your catalog of various floating point oddities

**Pull request merged**